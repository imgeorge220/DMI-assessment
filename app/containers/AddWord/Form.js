import styled from 'styled-components';

const Form = styled.form`
  margin-bottom: 1em;
  text-align: center;
`;

export default Form;
