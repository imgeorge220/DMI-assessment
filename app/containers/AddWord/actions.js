/*
 *
 * AddWord actions
 *
 */
import {
  CHANGE_NEW_WORD,
  ADD_WORD,
  ADD_WORD_SUCCESS,
  ADD_WORD_FAILURE,
} from './constants';

export function changeNewWord(newWord) {
  return {
    type: CHANGE_NEW_WORD,
    newWord,
  };
}

export function addWord() {
  return {
    type: ADD_WORD,
  };
}

export function wordAdded() {
  return {
    type: ADD_WORD_SUCCESS,
  };
}

export function wordAddError(error) {
  return {
    type: ADD_WORD_FAILURE,
    error,
  };
}
