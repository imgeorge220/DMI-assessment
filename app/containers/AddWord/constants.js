/*
 *
 * AddWord constants
 *
 */

export const CHANGE_NEW_WORD = 'app/AddWord/CHANGE_NEW_WORD';

export const ADD_WORD = 'app/AddWord/ADD_WORD';
export const ADD_WORD_SUCCESS = 'app/AddWord/ADD_WORD_SUCCESS';
export const ADD_WORD_FAILURE = 'app/AddWord/ADD_WORD_FAILURE';
