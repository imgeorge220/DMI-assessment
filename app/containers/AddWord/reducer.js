/*
 *
 * AddWord reducer
 *
 */
import produce from 'immer';
import {
  CHANGE_NEW_WORD,
  ADD_WORD,
  ADD_WORD_SUCCESS,
  ADD_WORD_FAILURE,
} from './constants';

export const initialState = {
  loading: false,
  error: false,
  success: false,
  newWord: '',
};

/* eslint-disable default-case, no-param-reassign */
const addWordReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_NEW_WORD:
        draft.newWord = action.newWord;
        break;

      case ADD_WORD:
        draft.loading = true;
        draft.error = false;
        draft.success = false;
        break;

      case ADD_WORD_SUCCESS:
        draft.loading = false;
        draft.success = true;
        draft.error = false;
        break;

      case ADD_WORD_FAILURE:
        draft.error = action.error;
        draft.success = false;
        draft.loading = false;
        break;
    }
  });

export default addWordReducer;
