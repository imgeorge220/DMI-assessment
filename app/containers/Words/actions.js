/*
 *
 * Words actions
 *
 */
import {
  LOAD_WORDS,
  LOAD_WORDS_SUCCESS,
  LOAD_WORDS_FAILURE,
  ADD_WORD,
  ADD_WORD_SUCCESS,
  ADD_WORD_FAILURE,
} from './constants';

export function loadWords() {
  return {
    type: LOAD_WORDS,
  };
}

export function wordsLoaded(words) {
  return {
    type: LOAD_WORDS_SUCCESS,
    words,
  };
}

export function wordsLoadError(error) {
  return {
    type: LOAD_WORDS_FAILURE,
    error,
  };
}

export function addWord() {
  return {
    type: ADD_WORD,
  };
}

export function wordAdded() {
  return {
    type: ADD_WORD_SUCCESS,
  };
}

export function wordAddError(error) {
  return {
    type: ADD_WORD_FAILURE,
    error,
  };
}
