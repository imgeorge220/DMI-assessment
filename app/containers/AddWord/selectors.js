import { createSelector } from 'reselect';
import { initialState } from './reducer';

/**
 * AddWords Selectors
 */

const selectAddWordState = state => state.addWord || initialState;

const makeSelectNewWord = () =>
  createSelector(
    selectAddWordState,
    substate => substate.newWord,
  );

const makeSelectNewWordLoading = () =>
  createSelector(
    selectAddWordState,
    substate => substate.loading,
  );

const makeSelectNewWordError = () =>
  createSelector(
    selectAddWordState,
    substate => substate.error,
  );

const makeSelectNewWordSuccess = () =>
  createSelector(
    selectAddWordState,
    substate => substate.success,
  );

export {
  selectAddWordState,
  makeSelectNewWord,
  makeSelectNewWordLoading,
  makeSelectNewWordError,
  makeSelectNewWordSuccess,
};
