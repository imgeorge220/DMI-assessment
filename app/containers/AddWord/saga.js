import { call, put, takeLatest, select } from 'redux-saga/effects';
import apiRequest from 'utils/apiRequest';
import { ADD_WORD } from './constants';
import { wordAdded, wordAddError } from './actions';
import { makeSelectNewWord } from './selectors';

/**
 * posts word to backend API
 */
export function* addWord() {
  const newWord = yield select(makeSelectNewWord());
  const requestURL = `http://localhost:3001`;
  const reqOptions = {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ word: newWord }),
  };

  try {
    // Call request helper (see 'utils/apiRequest')
    yield call(apiRequest, requestURL, reqOptions);
    yield put(wordAdded());
  } catch (err) {
    yield put(wordAddError(err));
  }
}

/**
 * Watcher Saga
 */
export default function* wordsWatcher() {
  yield takeLatest(ADD_WORD, addWord);
}
