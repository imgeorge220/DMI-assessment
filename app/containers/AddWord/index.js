/**
 *
 * AddWord
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import { useInjectSaga } from 'utils/injectSaga';
import { useInjectReducer } from 'utils/injectReducer';
import Form from './Form';
import Input from './Input';
import { changeNewWord } from './actions';
import { makeSelectNewWord } from './selectors';
import reducer from './reducer';
import saga from './saga';
import messages from './messages';

export function AddWord({ newWord, onChangeNewWord, onSubmitForm }) {
  useInjectReducer({ key: 'addWord', reducer });
  useInjectSaga({ key: 'addWord', saga });

  return (
    <div>
      <FormattedMessage {...messages.header} />
      <Form onSubmit={onSubmitForm}>
        <Input
          id="newWord"
          type="text"
          value={newWord}
          onChange={onChangeNewWord}
        />
      </Form>
    </div>
  );
}

AddWord.propTypes = {
  newWord: PropTypes.string,
  onChangeNewWord: PropTypes.func,
  onSubmitForm: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  newWord: makeSelectNewWord(),
});

function mapDispatchToProps(dispatch) {
  return {
    onChangeNewWord: evt => dispatch(changeNewWord(evt.target.value)),
    onSubmitForm: evt => {
      evt.preventDefault();
      // dispatch(addWord());
    }
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(withConnect)(AddWord);
