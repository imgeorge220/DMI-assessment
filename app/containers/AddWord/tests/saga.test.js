/**
 * Test sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { put, takeLatest } from 'redux-saga/effects';
import { ADD_WORD } from 'containers/AddWord/constants';
import { wordAdded, wordAddError } from 'containers/AddWord/actions';
import addWordsWatcher, { addWord } from '../saga';

describe('addWordSaga Saga', () => {
  let addWordGenerator;

  beforeEach(() => {
    addWordGenerator = addWord();

    const selectDescriptor = addWordGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = addWordGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the addWord action if it requests the data successfully', () => {
    const putDescriptor = addWordGenerator.next().value;
    expect(putDescriptor).toEqual(put(wordAdded()));
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const response = new Error('Test erros');
    const putDescriptor = addWordGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(wordAddError(response)));
  });

  describe('addWordsWatcher Saga', () => {
    const addWordsSaga = addWordsWatcher();

    it('should start task to watch for LOAD_REPOS action', () => {
      const takeLatestDescriptor = addWordsSaga.next().value;
      expect(takeLatestDescriptor).toEqual(takeLatest(ADD_WORD, addWord));
    });
  });
});
