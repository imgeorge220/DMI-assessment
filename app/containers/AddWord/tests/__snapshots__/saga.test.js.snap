// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addWordSaga Saga addWordsWatcher Saga should start task to watch for LOAD_REPOS action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addWordSaga Saga addWordsWatcher Saga should start task to watch for LOAD_REPOS action 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001",
      Object {
        "body": "{}",
        "headers": Object {
          "Accept": "application/json",
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`addWordSaga Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addWordSaga Saga should call the repoLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001",
      Object {
        "body": "{}",
        "headers": Object {
          "Accept": "application/json",
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`addWordSaga Saga should dispatch the addWord action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addWordSaga Saga should dispatch the addWord action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001",
      Object {
        "body": "{}",
        "headers": Object {
          "Accept": "application/json",
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
